src/IO/Observer/Actions/RegisterEvents.hpp

```diff
@@ -116,29 +116,36 @@ struct RegisterEventsWithObservers {
     std::vector<
         std::pair<observers::TypeOfObservation, observers::ObservationKey>>
         type_of_observation_and_observation_key_pairs;
-
-    const auto& triggers_and_events =
-        db::get<::Tags::EventsAndTriggersBase>(box);
-    for (const auto& trigger_and_events :
-         triggers_and_events.events_and_triggers()) {
-      for (const auto& event : trigger_and_events.second) {
-        if (auto obs_type_and_obs_key =
-                get_registration_observation_type_and_key(*event, box);
-            obs_type_and_obs_key.has_value()) {
-          type_of_observation_and_observation_key_pairs.push_back(
-              *obs_type_and_obs_key);
+    if constexpr (db::tag_is_retrievable_v<::Tags::EventsAndTriggersBase,
+                                           db::DataBox<DbTagList>>) {
+      const auto& triggers_and_events =
+          db::get<::Tags::EventsAndTriggersBase>(box);
+      for (const auto& trigger_and_events :
+           triggers_and_events.events_and_triggers()) {
+        for (const auto& event : trigger_and_events.second) {
+          if (auto obs_type_and_obs_key =
+                  get_registration_observation_type_and_key(*event, box);
+              obs_type_and_obs_key.has_value()) {
+            type_of_observation_and_observation_key_pairs.push_back(
+                *obs_type_and_obs_key);
+          }
         }
       }
-    }
 
-    for (const auto& [type_of_observation, observation_key] :
-         type_of_observation_and_observation_key_pairs) {
-      Parallel::simple_action<RegisterOrDeregisterAction>(
-          observer, observation_key,
-          observers::ArrayComponentId(
-              std::add_pointer_t<ParallelComponent>{nullptr},
-              Parallel::ArrayIndex<std::decay_t<ArrayIndex>>{array_index}),
-          type_of_observation);
+      for (const auto& [type_of_observation, observation_key] :
+           type_of_observation_and_observation_key_pairs) {
+        Parallel::simple_action<RegisterOrDeregisterAction>(
+            observer, observation_key,
+            observers::ArrayComponentId(
+                std::add_pointer_t<ParallelComponent>{nullptr},
+                Parallel::ArrayIndex<std::decay_t<ArrayIndex>>{array_index}),
+            type_of_observation);
+      }
+    } else {
+      ERROR(
+          "Cannot perform registration of events, "
+          "`::Tags::EventsAndTriggersBase` is not retrievable from the "
+          "DataBox");
```

Comment:
Yes, this function is instantiated for all databox types -- note it is `perform_registration_impl` now, not the `apply` because of the factoring to not repeat code in the register/deregister branches.
This is because the registration/deregistration has to happen during the pup. Without that, it is pretty hard to guarantee that we've accomplished the registration correctly before messages start getting sent again. However, the pup calls the registration functions much like the simple action pattern, so it has the same instantiation structure.